core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -662,6 +662,7 @@ private[spark] class TaskSetManager(
 
     val failureReason = s"Lost task ${info.id} in stage ${taskSet.id} (TID $tid, ${info.host}): " +
       reason.asInstanceOf[TaskFailedReason].toErrorString
+    var failureException: Option[Throwable] = None
     reason match {
```

Comment:
I'd prefer that you avoid the unnecessary mutability.  The `match` cases are all quite small with the exception of the `ExceptionFailure` case, which means that even with the `var` it's not all that readable or easy to figure out just where `failureException` is being set and that it is not being mutated elsewhere.  If you really want to shorten the `match` and make it more readable, I suggest pulling the logic of the `ExceptionFailure` case out into a helper function defined within `def handleFailedTask`, then make each of the cases in the `match` produce `None` or `Some(throwable)`.